File: js/src/cst-js.mjs
// JavaScript ↔ `.lino` CST converter (issue #138).
//
// Token-level lossless converter for JavaScript source. Produces a
// `lino-cst.js.*` flat CST. Round-trip is byte-faithful:
// `printJs(parseJs(src)) === src`.
//
// Scope: this converter operates at the token-stream level. It recognises:
//
//   - line comments (`// ...`)
//   - block comments (`/* ... */`)
//   - whitespace (spaces, tabs, CR, LF)
//   - hashbang on the first line (`#! ...`)
//   - string literals (`"..."`, `'...'`)
//   - template literals (``` `...${...}...` ```), with nested expressions
//   - numeric literals (decimal, hex, oct, bin, BigInt)
//   - regexp literals (best-effort, using a context heuristic)
//   - identifiers (Unicode XID-friendly)
//   - punctuation
//
// Regex disambiguation is the only non-trivial piece: a `/` is the start of a
// regex when it follows a token that cannot begin a binary-divide operand
// (e.g. `=`, `(`, `,`, `return`, etc.). We use the standard heuristic.

import { list, token, trivia, DIALECTS } from './cst.mjs';

const JS = DIALECTS.js;

/**
 * Parse JavaScript source into a `lino-cst.js.*` CST.
 *
 * @param {string} src JS source.
 * @returns {import('./cst.mjs').CstNode}
 */
export function parseJs(src) {
  return list(`${JS}.program`, tokeniseJs(String(src)));
}

/**
 * Print a JS CST back to source.
 * @param {import('./cst.mjs').CstNode} node
 * @returns {string}
 */
export function printJs(node) {
  const out = [];
  emit(node, out);
  return out.join('');
}

function emit(node, out) {
  if (!node) return;
  if (node.kind === 'token' || node.kind === 'trivia') {
    out.push(node.text);
    return;
  }
  if (node.kind === 'list') {
    if (node.open) out.push(node.open);
    for (const child of node.children) emit(child, out);
    if (node.close) out.push(node.close);
  }
}

const ID_START = /[A-Za-z_$]/;
const ID_CONT = /[A-Za-z0-9_$]/;
const DIGIT = /[0-9]/;
const HEX = /[0-9A-Fa-f]/;

const REGEX_PRECEDING_KEYWORDS = new Set([
  'return', 'typeof', 'instanceof', 'in', 'of', 'do', 'else', 'throw',
  'new', 'delete', 'void', 'await', 'yield', 'case',
]);

function tokeniseJs(src) {
  const out = [];
  let i = 0;
  let lastSignificant = null;

  if (src.startsWith('#!')) {
    let j = src.indexOf('\n');
    if (j === -1) j = src.length;
    out.push(trivia(src.substring(0, j), `${JS}.hashbang`));
    i = j;
  }

  while (i < src.length) {
    const c = src[i];

    if (c === ' ' || c === '\t' || c === '\r' || c === '\n') {
      let j = i;
      while (j < src.length && (src[j] === ' ' || src[j] === '\t' || src[j] === '\r' || src[j] === '\n')) j++;
      out.push(trivia(src.substring(i, j), `${JS}.whitespace`));
      i = j;
      continue;
    }

    if (c === '/' && src[i + 1] === '/') {
      let j = i + 2;
      while (j < src.length && src[j] !== '\n') j++;
      out.push(trivia(src.substring(i, j), `${JS}.comment.line`));
      i = j;
      continue;
    }

    if (c === '/' && src[i + 1] === '*') {
      const end = src.indexOf('*/', i + 2);
      const j = end === -1 ? src.length : end + 2;
      out.push(trivia(src.substring(i, j), `${JS}.comment.block`));
      i = j;
      continue;
    }

    if (c === '"' || c === "'") {
      const j = scanString(src, i + 1, c);
      const tok = token(src.substring(i, j), `${JS}.string_literal`);
      out.push(tok);
      lastSignificant = tok;
      i = j;
      continue;
    }

    if (c === '`') {
      const j = scanTemplate(src, i);
      const tok = token(src.substring(i, j), `${JS}.template_literal`);
      out.push(tok);
      lastSignificant = tok;
      i = j;
      continue;
    }

    if (c === '/' && canBeRegex(lastSignificant)) {
      const j = scanRegex(src, i);
      if (j > i + 1) {
        const tok = token(src.substring(i, j), `${JS}.regexp_literal`);
        out.push(tok);
        lastSignificant = tok;
        i = j;
        continue;
      }
    }

    if (DIGIT.test(c) || (c === '.' && DIGIT.test(src[i + 1]))) {
      const j = scanNumber(src, i);
      const tok = token(src.substring(i, j), `${JS}.numeric_literal`);
      out.push(tok);
      lastSignificant = tok;
      i = j;
      continue;
    }

    if (ID_START.test(c) || isUnicodeIdStart(c)) {
      let j = i + 1;
      while (j < src.length && (ID_CONT.test(src[j]) || isUnicodeIdContinue(src[j]))) j++;
      const tok = token(src.substring(i, j), `${JS}.ident`);
      out.push(tok);
      lastSignificant = tok;
      i = j;
      continue;
    }

    const tok = token(c, `${JS}.punct`);
    out.push(tok);
    lastSignificant = tok;
    i += 1;
  }

  return out;
}

function scanString(src, j, quote) {
  while (j < src.length) {
    const c = src[j];
    if (c === '\\') { j += 2; continue; }
    if (c === '\n' && (quote === '"' || quote === "'")) {
      // Unterminated; stop at end of line for safety.
      return j;
    }
    if (c === quote) return j + 1;
    j++;
  }
  return j;
}

function scanTemplate(src, i) {
  let j = i + 1;
  while (j < src.length) {
    const c = src[j];
    if (c === '\\') { j += 2; continue; }
    if (c === '`') return j + 1;
    if (c === '$' && src[j + 1] === '{') {
      // Skip nested ${ ... } expression with bracket counting.
      j += 2;
      let depth = 1;
      while (j < src.length && depth > 0) {
        const k = src[j];
        if (k === '{') depth++;
        else if (k === '}') depth--;
        else if (k === '"' || k === "'") j = scanString(src, j + 1, k) - 1;
        else if (k === '`') j = scanTemplate(src, j) - 1;
        else if (k === '/' && src[j + 1] === '/') {
          while (j < src.length && src[j] !== '\n') j++;
          continue;
        }
        else if (k === '/' && src[j + 1] === '*') {
          const e = src.indexOf('*/', j + 2);
          j = e === -1 ? src.length : e + 2;
          continue;
        }
        j++;
      }
      continue;
    }
    j++;
  }
  return j;
}

function scanRegex(src, i) {
  let j = i + 1;
  let inClass = false;
  while (j < src.length) {
    const c = src[j];
    if (c === '\\') { j += 2; continue; }
    if (c === '[') inClass = true;
    else if (c === ']') inClass = false;
    else if (c === '/' && !inClass) {
      j++;
      // Flags
      while (j < src.length && /[a-zA-Z]/.test(src[j])) j++;
      return j;
    } else if (c === '\n') {
      return i + 1;
    }
    j++;
  }
  return j;
}

function scanNumber(src, i) {
  let j = i;
  if (src[j] === '0' && (src[j + 1] === 'x' || src[j + 1] === 'X')) {
    j += 2;
    while (j < src.length && (HEX.test(src[j]) || src[j] === '_')) j++;
  } else if (src[j] === '0' && (src[j + 1] === 'o' || src[j + 1] === 'O')) {
    j += 2;
    while (j < src.length && /[0-7_]/.test(src[j])) j++;
  } else if (src[j] === '0' && (src[j + 1] === 'b' || src[j + 1] === 'B')) {
    j += 2;
    while (j < src.length && /[01_]/.test(src[j])) j++;
  } else {
    while (j < src.length && (DIGIT.test(src[j]) || src[j] === '_')) j++;
    if (src[j] === '.') {
      j++;
      while (j < src.length && (DIGIT.test(src[j]) || src[j] === '_')) j++;
    }
    if (src[j] === 'e' || src[j] === 'E') {
      j++;
      if (src[j] === '+' || src[j] === '-') j++;
      while (j < src.length && (DIGIT.test(src[j]) || src[j] === '_')) j++;
    }
  }
  if (src[j] === 'n') j++; // BigInt suffix
  return j;
}

function canBeRegex(prev) {
  if (!prev) return true;
  if (prev.kind === 'trivia') return true;
  if (prev.tag === `${JS}.ident`) {
    return REGEX_PRECEDING_KEYWORDS.has(prev.text);
  }
  if (prev.tag === `${JS}.punct`) {
    // After most punctuation, `/` starts a regex.
    return !')]'.includes(prev.text);
  }
  return false;
}

function isUnicodeIdStart(c) {
  if (!c) return false;
  return c.codePointAt(0) > 0x7f;
}

function isUnicodeIdContinue(c) {
  return isUnicodeIdStart(c);
}

File: rust/src/cst_js.rs
//! JavaScript ↔ `.lino` CST converter (issue #138).
//!
//! Token-level lossless converter for JavaScript source. Produces a
//! `lino-cst.js.*` flat CST whose round-trip is byte-faithful:
//! `print_js(&parse_js(src)) == src`. Mirrors `js/src/cst-js.mjs`
//! line for line.

use crate::cst::{dialects::JS, print_cst, CstNode};

/// Parse JavaScript source into a `lino-cst.js.*` CST.
pub fn parse_js(src: &str) -> CstNode {
    let children = tokenise(src);
    CstNode::list(format!("{}.program", JS), children)
}

/// Print a JS CST back to source.
pub fn print_js(node: &CstNode) -> String {
    print_cst(node)
}

#[derive(Clone, Copy)]
enum LastKind {
    None,
    Trivia,
    Ident,
    Punct,
    Other,
}

fn tokenise(src: &str) -> Vec<CstNode> {
    let chars: Vec<char> = src.chars().collect();
    let mut out: Vec<CstNode> = Vec::new();
    let mut i = 0usize;
    let mut last_kind = LastKind::None;
    let mut last_text: String = String::new();

    if chars.len() >= 2 && chars[0] == '#' && chars[1] == '!' {
        let mut j = 0usize;
        while j < chars.len() && chars[j] != '\n' {
            j += 1;
        }
        out.push(CstNode::trivia(
            chars[..j].iter().collect::<String>(),
            Some(&format!("{}.hashbang", JS)),
        ));
        i = j;
    }

    while i < chars.len() {
        let c = chars[i];

        if c == ' ' || c == '\t' || c == '\r' || c == '\n' {
            let mut j = i;
            while j < chars.len()
                && (chars[j] == ' ' || chars[j] == '\t' || chars[j] == '\r' || chars[j] == '\n')
            {
                j += 1;
            }
            out.push(CstNode::trivia(
                chars[i..j].iter().collect::<String>(),
                Some(&format!("{}.whitespace", JS)),
            ));
            i = j;
            last_kind = LastKind::Trivia;
            continue;
        }

        if c == '/' && chars.get(i + 1) == Some(&'/') {
            let mut j = i + 2;
            while j < chars.len() && chars[j] != '\n' {
                j += 1;
            }
            out.push(CstNode::trivia(
                chars[i..j].iter().collect::<String>(),
                Some(&format!("{}.comment.line", JS)),
            ));
            i = j;
            last_kind = LastKind::Trivia;
            continue;
        }

        if c == '/' && chars.get(i + 1) == Some(&'*') {
            let j = scan_block_comment(&chars, i);
            out.push(CstNode::trivia(
                chars[i..j].iter().collect::<String>(),
                Some(&format!("{}.comment.block", JS)),
            ));
            i = j;
            last_kind = LastKind::Trivia;
            continue;
        }

        if c == '"' || c == '\'' {
            let j = scan_string(&chars, i + 1, c);
            let text: String = chars[i..j].iter().collect();
            out.push(CstNode::token(
                text.clone(),
                Some(&format!("{}.string_literal", JS)),
            ));
            i = j;
            last_kind = LastKind::Other;
            last_text = text;
            continue;
        }

        if c == '`' {
            let j = scan_template(&chars, i);
            let text: String = chars[i..j].iter().collect();
            out.push(CstNode::token(
                text.clone(),
                Some(&format!("{}.template_literal", JS)),
            ));
            i = j;
            last_kind = LastKind::Other;
            last_text = text;
            continue;
        }

        if c == '/' && can_be_regex(last_kind, &last_text) {
            let j = scan_regex(&chars, i);
            if j > i + 1 {
                let text: String = chars[i..j].iter().collect();
                out.push(CstNode::token(
                    text.clone(),
                    Some(&format!("{}.regexp_literal", JS)),
                ));
                i = j;
                last_kind = LastKind::Other;
                last_text = text;
                continue;
            }
        }

        if c.is_ascii_digit()
            || (c == '.' && chars.get(i + 1).map(|x| x.is_ascii_digit()).unwrap_or(false))
        {
            let j = scan_number(&chars, i);
            let text: String = chars[i..j].iter().collect();
            out.push(CstNode::token(
                text.clone(),
                Some(&format!("{}.numeric_literal", JS)),
            ));
            i = j;
            last_kind = LastKind::Other;
            last_text = text;
            continue;
        }

        if is_ident_start(c) {
            let mut j = i + 1;
            while j < chars.len() && is_ident_continue(chars[j]) {
                j += 1;
            }
            let text: String = chars[i..j].iter().collect();
            out.push(CstNode::token(
                text.clone(),
                Some(&format!("{}.ident", JS)),
            ));
            i = j;
            last_kind = LastKind::Ident;
            last_text = text;
            continue;
        }

        let text = c.to_string();
        out.push(CstNode::token(
            text.clone(),
            Some(&format!("{}.punct", JS)),
        ));
        i += 1;
        last_kind = LastKind::Punct;
        last_text = text;
    }

    out
}

fn scan_block_comment(chars: &[char], i: usize) -> usize {
    let mut j = i + 2;
    while j < chars.len() {
        if chars[j] == '*' && chars.get(j + 1) == Some(&'/') {
            return j + 2;
        }
        j += 1;
    }
    chars.len()
}

fn scan_string(chars: &[char], mut j: usize, quote: char) -> usize {
    while j < chars.len() {
        let c = chars[j];
        if c == '\\' {
            j += 2;
            continue;
        }
        if c == '\n' && (quote == '"' || quote == '\'') {
            return j;
        }
        if c == quote {
            return j + 1;
        }
        j += 1;
    }
    j
}

fn scan_template(chars: &[char], i: usize) -> usize {
    let mut j = i + 1;
    while j < chars.len() {
        let c = chars[j];
        if c == '\\' {
            j += 2;
            continue;
        }
        if c == '`' {
            return j + 1;
        }
        if c == '$' && chars.get(j + 1) == Some(&'{') {
            j += 2;
            let mut depth = 1;
            while j < chars.len() && depth > 0 {
                let k = chars[j];
                if k == '{' {
                    depth += 1;
                } else if k == '}' {
                    depth -= 1;
                } else if k == '"' || k == '\'' {
                    j = scan_string(chars, j + 1, k).saturating_sub(1);
                } else if k == '`' {
                    j = scan_template(chars, j).saturating_sub(1);
                } else if k == '/' && chars.get(j + 1) == Some(&'/') {
                    while j < chars.len() && chars[j] != '\n' {
                        j += 1;
                    }
                    continue;
                } else if k == '/' && chars.get(j + 1) == Some(&'*') {
                    j = scan_block_comment(chars, j);
                    continue;
                }
                j += 1;
            }
            continue;
        }
        j += 1;
    }
    j
}

fn scan_regex(chars: &[char], i: usize) -> usize {
    let mut j = i + 1;
    let mut in_class = false;
    while j < chars.len() {
        let c = chars[j];
        if c == '\\' {
            j += 2;
            continue;
        }
        if c == '[' {
            in_class = true;
        } else if c == ']' {
            in_class = false;
        } else if c == '/' && !in_class {
            j += 1;
            while j < chars.len() && chars[j].is_ascii_alphabetic() {
                j += 1;
            }
            return j;
        } else if c == '\n' {
            return i + 1;
        }
        j += 1;
    }
    j
}

fn scan_number(chars: &[char], i: usize) -> usize {
    let mut j = i;
    if chars.get(j) == Some(&'0') && matches!(chars.get(j + 1), Some('x') | Some('X')) {
        j += 2;
        while j < chars.len() && (chars[j].is_ascii_hexdigit() || chars[j] == '_') {
            j += 1;
        }
    } else if chars.get(j) == Some(&'0') && matches!(chars.get(j + 1), Some('o') | Some('O')) {
        j += 2;
        while j < chars.len() && matches!(chars[j], '0'..='7' | '_') {
            j += 1;
        }
    } else if chars.get(j) == Some(&'0') && matches!(chars.get(j + 1), Some('b') | Some('B')) {
        j += 2;
        while j < chars.len() && matches!(chars[j], '0' | '1' | '_') {
            j += 1;
        }
    } else {
        while j < chars.len() && (chars[j].is_ascii_digit() || chars[j] == '_') {
            j += 1;
        }
        if chars.get(j) == Some(&'.') {
            j += 1;
            while j < chars.len() && (chars[j].is_ascii_digit() || chars[j] == '_') {
                j += 1;
            }
        }
        if matches!(chars.get(j), Some('e') | Some('E')) {
            j += 1;
            if matches!(chars.get(j), Some('+') | Some('-')) {
                j += 1;
            }
            while j < chars.len() && (chars[j].is_ascii_digit() || chars[j] == '_') {
                j += 1;
            }
        }
    }
    if chars.get(j) == Some(&'n') {
        j += 1;
    }
    j
}

const REGEX_PRECEDING_KEYWORDS: &[&str] = &[
    "return",
    "typeof",
    "instanceof",
    "in",
    "of",
    "do",
    "else",
    "throw",
    "new",
    "delete",
    "void",
    "await",
    "yield",
    "case",
];

fn can_be_regex(last_kind: LastKind, last_text: &str) -> bool {
    match last_kind {
        LastKind::None | LastKind::Trivia => true,
        LastKind::Ident => REGEX_PRECEDING_KEYWORDS.iter().any(|k| *k == last_text),
        LastKind::Punct => !(last_text == ")" || last_text == "]"),
        LastKind::Other => false,
    }
}

fn is_ident_start(c: char) -> bool {
    c == '_' || c == '$' || c.is_ascii_alphabetic() || (c as u32) > 0x7F
}

fn is_ident_continue(c: char) -> bool {
    c == '_' || c == '$' || c.is_ascii_alphanumeric() || (c as u32) > 0x7F
}

#[cfg(test)]
mod tests {
    use super::*;

    fn rt(src: &str) {
        let node = parse_js(src);
        let back = print_js(&node);
        assert_eq!(back, src, "round-trip mismatch");
    }

    #[test]
    fn empty_string() {
        rt("");
    }

    #[test]
    fn simple_const() {
        rt("const x = 1;\n");
    }

    #[test]
    fn comments_and_strings() {
        rt("// hi\n/* block */ const s = 'a\\'b';\n");
    }

    #[test]
    fn template_with_nested_expr() {
        rt("const t = `nested ${`inner ${1+2}`} done`;\n");
    }

    #[test]
    fn regex_after_assignment() {
        rt("const r = /foo\\/bar/g;\n");
    }

    #[test]
    fn divide_after_paren_is_not_regex() {
        rt("const x = (1)/2;\n");
    }

    #[test]
    fn bigint_and_hex() {
        rt("const n = 0xff_ff;\nconst b = 0b1010n;\nconst f = 3.14e-2;\n");
    }

    #[test]
    fn hashbang_preserved() {
        rt("#!/usr/bin/env node\nconsole.log(\"hi\");\n");
    }
}
